File: leetCode/python/p38_CountAndSay/lc38.py
#!/usr/bin/python3

class Solution:
    '''
        The count-and-say sequence is a sequence of digit strings defined by the recursive formula:
        countAndSay(1) = "1"
        countAndSay(n) is the way you would "say" the digit string from countAndSay(n-1), 
        which is then converted into a different digit string.

        To determine how you "say" a digit string, 
        split it into the minimal number of substrings such that each substring 
        contains exactly one unique digit. Then for each substring, 
        say the number of digits, then say the digit. 
        Finally, concatenate every said digit.
    '''
    def countAndSay(self, n: int) -> str:
        res = '1'

        for i in range(1, n):
            count = 0
            prev = '.'
            curr_str = ''

            for char in res:
                if char == prev or prev == '.':
                    count += 1
                else:
                    curr_str += str(count) + prev
                    count = 1
                prev = char 

            curr_str += str(count) + prev
            res = curr_str 

        return res

if __name__ == '__main__':

    s = Solution()

    # Test 1:

    print('Test 1:')
    n1 = 1
    sol1 = s.countAndSay(n1)

    print('Number: ' + str(n1))
    
    print('Solution: ' + str(sol1))

    # Test 2:
       
    print('\nTest 2:')
    n2 = 4
    sol2 = s.countAndSay(n2)

    print('Number: ' + str(n2))
    
    print('Solution: ' + str(sol2))

File: leetCode/cpp/p38_CountAndSay/lc38.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    string countAndSay(int n);
};

string Solution::countAndSay(int n) {

    string res = "1";

    for(int i=1; i<n; i++) {

        int count = 0;
        char prev = '.';
        string curr_str = "";

        for(char ch : res) {
            if(ch == prev || prev == '.')
                count++;
            else {
                curr_str += to_string(count) + prev;        
                count = 1;
            }
            prev = ch;
        }
        
        curr_str += to_string(count) + prev;
        res = curr_str;
    }
 
    return res;    
}

int main(void) {

    Solution s;

    /* Test 1: */ 
    cout << "Test 1:\n";

    int n1 = 1;

    cout << "Input 1: " << n1 << '\n';

    cout << "Solution 1: " << s.countAndSay(n1) << '\n';

    /* Test 2: */ 
    cout << "\nTest 2:\n";

    int n2 = 4;

    cout << "Input 2: " << n2 << '\n';

    cout << "Solution 2: " << s.countAndSay(n2) << '\n';

    return EXIT_SUCCESS;
}
